### PR TITLE
Create MLJ standard docstrings for all Nearest Neighbor models

### PR DESCRIPTION
**Pull request by OkonSamuel**
Addresses #51

### CODE REVIEW

File: src/models.jl [hunk 1]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+Here:
+
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- `y` is the target, which can be any `AbstractVector` whose element scitype is
+  `<:Finite`; check the scitype with `scitype(y)`
+
+- `w` is the observation weights which can either be `nothing`(default) or an 
+  `AbstractVector` whoose element scitype is `Count` or `Continuous`. This is
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  `AbstractVector` whose element scitype is `Count` or `Continuous`. This is 
```

File: src/models.jl [hunk 2]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- `y` is the target, which can be any `AbstractVector` whose element scitype is
+  `<:Finite`; check the scitype with `scitype(y)`
+
+- `w` is the observation weights which can either be `nothing`(default) or an 
+  `AbstractVector` whoose element scitype is `Count` or `Continuous`. This is 
+  different from `weights` kernel which is an hyperparameter to the model, see below.
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  different from `weights` kernel which is a model hyperparameter, see below.
```

File: src/models.jl [hunk 3]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+
+# Fitted parameters
+
+$KNNFITTEDPARAMS
+
+# Examples
+```
+using MLJ
+KNNClassifier = @load KNNClassifier pkg=NearestNeighborModels
+X, y = @load_crabs; # loads the crabs dataset from MLJBase
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
X, y = @load_crabs; # a table and a vector from the crabs dataset
```

File: src/models.jl [hunk 4]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+OR
+    mach = machine(model, X, y, w)
+    
+Here:
+
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- y` is the target, which can be any table of responses whose element scitype is either
+  `<:Finite`; check the scitype with `scitype(y)`.
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  `<:Finite` (`<:Multiclass` or `<:OrderedFactor` will do); check the columns scitypes with `schema(y)`. 
```

File: src/models.jl [hunk 5]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- y` is the target, which can be any table of responses whose element scitype is either
+  `<:Finite`; check the scitype with `scitype(y)`. 
+  Each column of `y` is assumed to belong to a common categorical pool.  
+
+- `w` is the observation weights which can either be `nothing`(default) or an 
+  `AbstractVector` whoose element scitype is `Count` or `Continuous`. This is different
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  `AbstractVector` whose element scitype is `Count` or `Continuous`. This is different 
```

File: src/models.jl [hunk 6]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- y` is the target, which can be any table of responses whose element scitype is either
+  `<:Finite`; check the scitype with `scitype(y)`. 
+  Each column of `y` is assumed to belong to a common categorical pool.  
+
+- `w` is the observation weights which can either be `nothing`(default) or an 
+  `AbstractVector` whoose element scitype is `Count` or `Continuous`. This is different 
+  from `weights` kernel which is an hyperparameter to the model, see below.
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  from `weights` kernel which is a model hyperparameter, see below.
```

File: src/models.jl [hunk 7]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+    (`ColumnTable`, `DictTable`). The type of table type to use for predictions.
+    Setting to `ColumnTable` might improve performance for narrow tables while setting to 
+    `DictTable` improves performance for wide tables.
+
+# Operations
+
+- `predict(mach, Xnew)`: Return predictions of the target given features `Xnew`, which
+  should have same scitype as `X` above. Predictions are either a `ColumnTable` or 
+  `DictTable` of `UnivariateFiniteVector` columns depending on the value set for the 
+  `output_type` parameter discussed above.
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  `output_type` parameter discussed above. The probabilistic predictions are uncalibrated. 
```

File: src/models.jl [hunk 8]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+See also [`MultitargetKNNRegressor`](@ref)
+"""
+KNNRegressor
+
+"""
+$(MMI.doc_header(MultitargetKNNRegressor))
+
+Multi-target K-Nearest Neighbors regressor (MultitargetKNNRegressor) is a variation of 
+[`KNNRegressor](@ref) that assumes the target variable is vector-valued with
+`Continuous` components.
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
`Continuous` components. (Target data must be presented as a table, however.)
```

File: src/models.jl [hunk 9]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+See also [`MultitargetKNNClassifier`](@ref)
+"""
+KNNClassifier
+
+"""
+$(MMI.doc_header(MultitargetKNNClassifier))
+
+Multi-target K-Nearest Neighbors Classifier (MultitargetKNNClassifier) is a variation of 
+[`KNNClassifier](@ref) that assumes the target variable is vector-valued with
+`Multiclass` or `OrderedFactor` components.
```
Comment (ablaom) [inline src/models.jl]: Probably we should mitigate possible confusion with requirements for `y` below. Something like:

```suggestion
`Multiclass` or `OrderedFactor` components. (Target data must be presented as a table, however.)
```

File: src/models.jl [hunk 10]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+OR
+    mach = machine(model, X, y, w)
+
+Here:
+
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- `y` is the target, which can be any table of responses whose element scitype is 
+  `Continuous`; check the scitype with `scitype(y)`.
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  `Continuous`; check column scitypes with `schema(y)`.
```

File: src/models.jl [hunk 11]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+OR
+    mach = machine(model, X, y, w)
+
+Here:
+
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- `y` is the target, which can be any `AbstractVector` whose element scitype is
+  `<:Finite`; check the scitype with `scitype(y)`
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
  `<:Finite` (`<:Multiclass` or `<:OrderedFactor` will do); check the scitype with `scitype(y)`
```

File: src/models.jl [hunk 12]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+with a new point by taking a vote over the classes of the k-nearest points. Each neighbor 
+vote is assigned a weight based on proximity of the neighbor point to the test point 
+according to a specified distance metric.
+
+For more information about the weighting kernels, see the paper by Geler et.al 
+[Comparison of different weighting schemes for the kNN classifier on time-series data](https://perun.pmf.uns.ac.rs/radovanovic/publications/2016-kais-knn-weighting.pdf). 
+
+# Training data
+In MLJ or MLJBase, bind an instance `model` to data with
+    mach = machine(model, X, y)
```
Comment (ablaom) [inline src/models.jl:551]: For some reason the code fencing is not kicking in here. When I view the docstring from the REPL, I'm getting

```
In MLJ or MLJBase, bind an instance model to data with mach = machine(model, X, y) OR mach = machine(model,
  X, y, w)
```

Perhaps you need spare lines around the code or you have tab character instead of 4 spaces???

File: src/models.jl [hunk 13]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+
+Here:
+
+- `X` is any table of input features (eg, a `DataFrame`) whose columns are of scitype
+  `Continuous`; check column scitypes with `schema(X)`.
+
+- `y` is the target, which can be any `AbstractVector` whose element scitype is
+  `<:Finite`; check the scitype with `scitype(y)`
+
+- `w` is the observation weights which can either be `nothing`(default) or an
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
- `w` is the observation weights which can either be `nothing` (default) or an 
```

File: src/models.jl [hunk 14]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+    * `reorder::Bool = true` : if `true` then points which are close in 
+        distance are placed close in memory. In this case, a copy of the original data 
+        will be made so that the original data is left unmodified. Setting this to `true` 
+        can significantly improve performance of the specified `algorithm` 
+        (except `:brutetree`). This option is ignored and always taken as `false` for 
+        `algorithm = :brutetree`.
+    * `weights::KNNKernel=Uniform()` : kernel used in assigning weights to the 
+        k-nearest neighbors for each observation. An instance of one of the types in 
+        `list_kernels()`. User-defined weighting functions can be passed by wrapping the 
+        function in a [`UserDefinedKernel`](@ref) kernel (see ?UserDefinedKernel for more
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
        function in a [`UserDefinedKernel`](@ref) kernel (do `?NearestNeighborModels.UserDefinedKernel` for more 
```

File: src/models.jl [hunk 15]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+```
+See also [`MultitargetKNNClassifier`](@ref)
+"""
+KNNClassifier
+
+"""
+$(MMI.doc_header(MultitargetKNNClassifier))
+
+Multi-target K-Nearest Neighbors Classifier (MultitargetKNNClassifier) is a variation of 
+[`KNNClassifier](@ref) that assumes the target variable is vector-valued with
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
[`KNNClassifier`](@ref) that assumes the target variable is vector-valued with
```

File: src/models.jl [hunk 16]
```diff
@@ -496,6 +473,320 @@ metadata_model(
...
+```
+See also [`MultitargetKNNRegressor`](@ref)
+"""
+KNNRegressor
+
+"""
+$(MMI.doc_header(MultitargetKNNRegressor))
+
+Multi-target K-Nearest Neighbors regressor (MultitargetKNNRegressor) is a variation of 
+[`KNNRegressor](@ref) that assumes the target variable is vector-valued with
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
[`KNNRegressor`](@ref) that assumes the target variable is vector-valued with
```

**Review by ablaom — COMMENTED** [review]
Great to have this @OkonSamuel .

Apart from the comments above:

- [x]  We need to move the docstrings so that they occur *after* the metadata declaration because the docstring header interpolator is not working:

```
A model type for constructing a multitarget knn regressor, based on unknown.jl (unknown), and implementing
  the MLJ model interface.
```

- [x]  I suggest we declare `human_name` traits, eg
```human_name = "K-nearest neighbor classifier"``` to make the doc header read better.

File: src/models.jl [hunk 17]
```diff
@@ -468,34 +451,374 @@ metadata_model(
...
     weights = true,
-    descr = KNNClassifierDescription,
     path = "$(PKG).KNNClassifier"
 )
 
 metadata_model(
     MultitargetKNNClassifier,
     input = Table(Continuous),
     target = Table(Finite),
+    human_name = "Multitarget K-nearest neighbor classifier",
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
    human_name = "multitarget K-nearest neighbor classifier",
```

File: src/models.jl [hunk 18]
```diff
@@ -468,34 +451,374 @@ metadata_model(
...
     weights = true,
-    descr = KNNClassifierDescription,
     path = "$(PKG).MultitargetKNNClassifier"
 )
 
 metadata_model(
     MultitargetKNNRegressor,
     input = Table(Continuous),
     target = Table(Continuous),
+    human_name = "Multitarget K-nearest neighbor regressor",
```
Comment (ablaom) [inline src/models.jl]:
```suggestion
    human_name = "multitarget K-nearest neighbor regressor",
```

**Review by ablaom — APPROVED** [review]
Thanks again for this.
